File: mini-nn-python/mini_nn/trainer.py
"""Training utilities for neural networks."""

import numpy as np
from dataclasses import dataclass, field
from typing import List, Optional
from .tensor import Tensor
from .network import Network
from .loss import Loss, LossType
from .optimizer import Optimizer


@dataclass
class TrainingConfig:
    """Configuration for training."""
    epochs: int = 100
    batch_size: int = 32
    validation_split: float = 0.2
    shuffle: bool = True
    verbose: bool = True
    early_stop_patience: int = 10


@dataclass
class TrainingHistory:
    """Training history container."""
    train_loss: List[float] = field(default_factory=list)
    train_accuracy: List[float] = field(default_factory=list)
    val_loss: List[float] = field(default_factory=list)
    val_accuracy: List[float] = field(default_factory=list)


class Trainer:
    """Neural network trainer."""

    def __init__(
        self,
        config: TrainingConfig,
        rng: Optional[np.random.Generator] = None,
    ):
        """Initialize trainer."""
        self.config = config
        self.rng = rng if rng is not None else np.random.default_rng()

    def fit(
        self,
        network: Network,
        x: Tensor,
        y: Tensor,
        loss_type: LossType,
        optimizer: Optimizer,
    ) -> TrainingHistory:
        """Train the network."""
        history = TrainingHistory()
        
        n_samples = x.rows
        n_val = int(n_samples * self.config.validation_split)
        n_train = n_samples - n_val
        
        # Shuffle and split data
        indices = np.arange(n_samples)
        self.rng.shuffle(indices)
        
        train_idx = indices[:n_train]
        val_idx = indices[n_train:]
        
        x_train = self._select_rows(x, train_idx)
        y_train = self._select_rows(y, train_idx)
        x_val = self._select_rows(x, val_idx)
        y_val = self._select_rows(y, val_idx)
        
        best_val_loss = float("inf")
        patience_counter = 0
        
        for epoch in range(self.config.epochs):
            # Shuffle training data
            if self.config.shuffle:
                perm = self.rng.permutation(n_train)
                x_train = self._select_rows(x_train, perm)
                y_train = self._select_rows(y_train, perm)
            
            # Training
            train_loss, train_acc = self._train_epoch(
                network, x_train, y_train, loss_type, optimizer
            )
            
            # Validation
            val_loss, val_acc = self._evaluate(network, x_val, y_val, loss_type)
            
            history.train_loss.append(train_loss)
            history.train_accuracy.append(train_acc)
            history.val_loss.append(val_loss)
            history.val_accuracy.append(val_acc)
            
            if self.config.verbose and (epoch + 1) % 10 == 0:
                print(
                    f"Epoch {epoch + 1:3d}: "
                    f"train_loss={train_loss:.4f}, train_acc={train_acc * 100:.2f}%, "
                    f"val_loss={val_loss:.4f}, val_acc={val_acc * 100:.2f}%"
                )
            
            # Early stopping
            if val_loss < best_val_loss:
                best_val_loss = val_loss
                patience_counter = 0
            else:
                patience_counter += 1
                if patience_counter >= self.config.early_stop_patience:
                    if self.config.verbose:
                        print(f"Early stopping at epoch {epoch + 1}")
                    break
        
        return history

    def _train_epoch(
        self,
        network: Network,
        x: Tensor,
        y: Tensor,
        loss_type: LossType,
        optimizer: Optimizer,
    ) -> tuple:
        """Run one training epoch."""
        n_samples = x.rows
        batch_size = self.config.batch_size
        n_batches = (n_samples + batch_size - 1) // batch_size
        
        total_loss = 0.0
        correct = 0
        total = 0
        
        for b in range(n_batches):
            start = b * batch_size
            end = min(start + batch_size, n_samples)
            
            x_batch = x.slice_rows(start, end)
            y_batch = y.slice_rows(start, end)
            
            # Forward pass
            pred = network.forward(x_batch)
            
            # Compute loss
            batch_loss = Loss.compute(pred, y_batch, loss_type)
            total_loss += batch_loss * (end - start)
            
            # Compute accuracy
            correct += self._compute_accuracy(pred, y_batch)
            total += end - start
            
            # Backward pass
            grad = Loss.gradient(pred, y_batch, loss_type)
            network.backward(grad)
            
            # Update weights
            optimizer.step(network.get_layers())
        
        return total_loss / n_samples, correct / total

    def _evaluate(
        self,
        network: Network,
        x: Tensor,
        y: Tensor,
        loss_type: LossType,
    ) -> tuple:
        """Evaluate network on data."""
        pred = network.forward(x)
        loss = Loss.compute(pred, y, loss_type)
        correct = self._compute_accuracy(pred, y)
        return loss, correct / x.rows

    def _compute_accuracy(self, pred: Tensor, target: Tensor) -> int:
        """Compute classification accuracy."""
        rows, cols = pred.shape
        correct = 0
        
        pred_data = pred.numpy()
        target_data = target.numpy()
        
        for i in range(rows):
            if cols == 1:
                # Binary classification
                pred_class = 1 if pred_data[i, 0] >= 0.5 else 0
                target_class = int(round(target_data[i, 0]))
                if pred_class == target_class:
                    correct += 1
            else:
                # Multi-class classification
                pred_argmax = np.argmax(pred_data[i])
                target_argmax = np.argmax(target_data[i])
                if pred_argmax == target_argmax:
                    correct += 1
        
        return correct

    def _select_rows(self, t: Tensor, indices: np.ndarray) -> Tensor:
        """Select rows by indices."""
        return Tensor(t.numpy()[indices])

File: mini-nn-java/src/main/java/com/mininn/Trainer.java
package com.mininn;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

/**
 * Training utilities for neural networks.
 */
public class Trainer {
    private final int epochs;
    private final int batchSize;
    private final double validationSplit;
    private final boolean shuffle;
    private final boolean verbose;
    private final int earlyStopPatience;
    private final Random rng;

    public Trainer(int epochs, int batchSize, double validationSplit, 
                   boolean shuffle, boolean verbose, int earlyStopPatience, Random rng) {
        this.epochs = epochs;
        this.batchSize = batchSize;
        this.validationSplit = validationSplit;
        this.shuffle = shuffle;
        this.verbose = verbose;
        this.earlyStopPatience = earlyStopPatience;
        this.rng = rng;
    }

    public TrainingHistory fit(Network network, Tensor x, Tensor y, Loss loss, Optimizer optimizer) {
        TrainingHistory history = new TrainingHistory();

        int nSamples = x.getRows();
        int nVal = (int) (nSamples * validationSplit);
        int nTrain = nSamples - nVal;

        // Create shuffled indices
        List<Integer> indices = new ArrayList<>();
        for (int i = 0; i < nSamples; i++) {
            indices.add(i);
        }
        Collections.shuffle(indices, rng);

        // Split data
        Tensor xTrain = selectRows(x, indices.subList(0, nTrain));
        Tensor yTrain = selectRows(y, indices.subList(0, nTrain));
        Tensor xVal = selectRows(x, indices.subList(nTrain, nSamples));
        Tensor yVal = selectRows(y, indices.subList(nTrain, nSamples));

        double bestValLoss = Double.POSITIVE_INFINITY;
        int patienceCounter = 0;

        for (int epoch = 0; epoch < epochs; epoch++) {
            // Shuffle training data
            if (shuffle) {
                shuffleData(xTrain, yTrain);
            }

            // Training
            double[] trainMetrics = trainEpoch(network, xTrain, yTrain, loss, optimizer);
            double trainLoss = trainMetrics[0];
            double trainAcc = trainMetrics[1];

            // Validation
            double[] valMetrics = evaluate(network, xVal, yVal, loss);
            double valLoss = valMetrics[0];
            double valAcc = valMetrics[1];

            history.addEpoch(trainLoss, trainAcc, valLoss, valAcc);

            if (verbose && (epoch + 1) % 10 == 0) {
                System.out.printf("Epoch %3d: train_loss=%.4f, train_acc=%.2f%%, val_loss=%.4f, val_acc=%.2f%%%n",
                    epoch + 1, trainLoss, trainAcc * 100, valLoss, valAcc * 100);
            }

            // Early stopping
            if (valLoss < bestValLoss) {
                bestValLoss = valLoss;
                patienceCounter = 0;
            } else {
                patienceCounter++;
                if (patienceCounter >= earlyStopPatience) {
                    if (verbose) {
                        System.out.printf("Early stopping at epoch %d%n", epoch + 1);
                    }
                    break;
                }
            }
        }

        return history;
    }

    private double[] trainEpoch(Network network, Tensor x, Tensor y, Loss loss, Optimizer optimizer) {
        int nSamples = x.getRows();
        int nBatches = (nSamples + batchSize - 1) / batchSize;

        double totalLoss = 0;
        int correct = 0;
        int total = 0;

        for (int b = 0; b < nBatches; b++) {
            int start = b * batchSize;
            int end = Math.min(start + batchSize, nSamples);

            Tensor xBatch = x.sliceRows(start, end);
            Tensor yBatch = y.sliceRows(start, end);

            // Forward pass
            Tensor pred = network.forward(xBatch);

            // Compute loss
            double batchLoss = loss.compute(pred, yBatch);
            totalLoss += batchLoss * (end - start);

            // Compute accuracy
            correct += computeAccuracy(pred, yBatch);
            total += end - start;

            // Backward pass
            Tensor grad = loss.gradient(pred, yBatch);
            network.backward(grad);

            // Update weights
            optimizer.step(network.getLayers());
        }

        return new double[]{totalLoss / nSamples, (double) correct / total};
    }

    private double[] evaluate(Network network, Tensor x, Tensor y, Loss loss) {
        Tensor pred = network.forward(x);
        double lossValue = loss.compute(pred, y);
        int correct = computeAccuracy(pred, y);
        return new double[]{lossValue, (double) correct / x.getRows()};
    }

    private int computeAccuracy(Tensor pred, Tensor target) {
        int correct = 0;
        int rows = pred.getRows();
        int cols = pred.getCols();

        for (int i = 0; i < rows; i++) {
            if (cols == 1) {
                // Binary classification
                double p = pred.get(i, 0);
                double t = target.get(i, 0);
                int predClass = p >= 0.5 ? 1 : 0;
                int targetClass = (int) Math.round(t);
                if (predClass == targetClass) {
                    correct++;
                }
            } else {
                // Multi-class classification
                int predArgmax = argmax(pred.getRow(i));
                int targetArgmax = argmax(target.getRow(i));
                if (predArgmax == targetArgmax) {
                    correct++;
                }
            }
        }

        return correct;
    }

    private int argmax(double[] arr) {
        int maxIdx = 0;
        double maxVal = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > maxVal) {
                maxVal = arr[i];
                maxIdx = i;
            }
        }
        return maxIdx;
    }

    private Tensor selectRows(Tensor t, List<Integer> indices) {
        int cols = t.getCols();
        Tensor result = new Tensor(indices.size(), cols);
        for (int i = 0; i < indices.size(); i++) {
            for (int j = 0; j < cols; j++) {
                result.set(i, j, t.get(indices.get(i), j));
            }
        }
        return result;
    }

    private void shuffleData(Tensor x, Tensor y) {
        int n = x.getRows();
        int cx = x.getCols();
        int cy = y.getCols();

        for (int i = n - 1; i > 0; i--) {
            int j = rng.nextInt(i + 1);
            // Swap rows i and j
            for (int k = 0; k < cx; k++) {
                double tmp = x.get(i, k);
                x.set(i, k, x.get(j, k));
                x.set(j, k, tmp);
            }
            for (int k = 0; k < cy; k++) {
                double tmp = y.get(i, k);
                y.set(i, k, y.get(j, k));
                y.set(j, k, tmp);
            }
        }
    }

    /**
     * Training history container.
     */
    public static class TrainingHistory {
        private final List<Double> trainLoss = new ArrayList<>();
        private final List<Double> trainAccuracy = new ArrayList<>();
        private final List<Double> valLoss = new ArrayList<>();
        private final List<Double> valAccuracy = new ArrayList<>();

        public void addEpoch(double tLoss, double tAcc, double vLoss, double vAcc) {
            trainLoss.add(tLoss);
            trainAccuracy.add(tAcc);
            valLoss.add(vLoss);
            valAccuracy.add(vAcc);
        }

        public List<Double> getTrainLoss() { return trainLoss; }
        public List<Double> getTrainAccuracy() { return trainAccuracy; }
        public List<Double> getValLoss() { return valLoss; }
        public List<Double> getValAccuracy() { return valAccuracy; }
    }
}
